src/celeritas/io/RootStepWriter.cc

```diff
@@ -0,0 +1,247 @@
+// Copyright 2022 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/io/RootStepWriter.cc
+//---------------------------------------------------------------------------//
+#include "RootStepWriter.hh"
+
+#include <algorithm>
+#include <tuple>
+#include <TBranch.h>
+#include <TFile.h>
+#include <TTree.h>
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Construct writer with RootFileManager, ParticleParams (to convert particle
+ * id to pdg), and the selection of data to be tallied.
+ */
+RootStepWriter::RootStepWriter(SPRootFileManager io_manager,
+                               SPParticleParams  particle_params,
+                               StepSelection     selection,
+                               Filters           filters)
+    : StepInterface()
+    , root_manager_(io_manager)
+    , particles_(particle_params)
+    , selection_(selection)
+    , filters_(filters)
+{
+    CELER_EXPECT(root_manager_);
+    this->make_tree();
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Set the number of entries (i.e. number of steps) stored in memory before
+ * ROOT flushes the data to disk. Default is ~32MB of compressed data.
+ *
+ * See `SetAutoFlush` in ROOT TTree Class reference for details:
+ * https://root.cern.ch/doc/master/classTTree.html
+ */
+void RootStepWriter::set_auto_flush(long num_entries)
+{
+    CELER_EXPECT(root_manager_);
+    CELER_EXPECT(tstep_tree_);
+    tstep_tree_->SetAutoFlush(num_entries);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Collect step data from each track on each thread id.
+ */
+void RootStepWriter::execute(StateHostRef const& steps)
+{
+#define IF_SELECTED(ATTR, VAL) \
+    do                         \
+    {                          \
+        if (selection_.ATTR)   \
+        {                      \
+            tstep_.ATTR = VAL; \
+        }                      \
+    } while (0)
+
+#define IF_POINT_SELECTED(ATTR, VAL) \
+    do                               \
+    {                                \
+        if (selection_point.ATTR)    \
+        {                            \
+            tpoint.ATTR = VAL;       \
+        }                            \
+    } while (0)
+
+#define IF_POINT_REAL3_SELECTED(ATTR, TPOINT_ATTR)          \
+    do                                                      \
+    {                                                       \
+        if (selection_point.ATTR)                           \
+        {                                                   \
+            this->copy_real3(point.ATTR[tid], TPOINT_ATTR); \
+        }                                                   \
+    } while (0)
```

Comment:
I'm not sure these macros are justifiable (i.e. do they make the code less prone to typos and more readable, or do they make it harder to see what's happening and easier to incorrectly call code?). The macros themselves rely on a few local variable names such as `selection_point` and `tpoint`, aren't quite consistent (the REAL3 has `[tid]` and the others don't), and save only a single reuse of `ATTR` and one line of code (not including braces).

To make this cleaner and use a single macro, maybe you could define some helper functions:
```c++
template<class I, class T>
inline void copy_if_selected(bool option, const OpaqueId<I, T>& src, T* dst)
{
    if (option)
    {
        CELER_EXPECT(src);
        *dst = src.unchecked_get();
    }
}
```
for the different types (quantity, real3, ...) and then have your macro defined as:
```c++
#define RSW_COPY(ATTR, VALUE) \
    copy_if_selected(selection_.ATTR, steps.ATTR[tid], &tstep_.ATTR)
```
You can make this work even with the `StepPoint` by making your `tstep_` helper class use an EnumArray rather than an array: since you're not exporting that root interface it doesn't matter.

Reply:
My suggestion here can still help:
```c++
template<class T>
inline void copy_if_selected(bool option, const T& src, T* dst)
{
    if (option)
    {
        CELER_EXPECT(src);
        *dst = src.unchecked_get();
    }
}

/* add the opaqueid code to do 'get' */

inline void copy_if_selected(bool option, const Real3& src, std::array<double, 3>* dst)
{
    if (option)
    {
        CELER_EXPECT(src);
        *dst = src.unchecked_get();
    }
}
```

then you can just do
```c++
#define RSW_STORE(ATTR) copy_if_selected(selection_.ATTR, steps.ATTR[tid], tstep_.ATTR)
```

less error prone, shorter, easier to read.